File: cli/src/core/VscodeWorkspaceLocator.ts
/**
 * VscodeWorkspaceLocator
 *
 * Per-platform path resolution and workspace.json scanning for VS Code-family
 * user data directories. Used by both Cursor (`flavor: "Cursor"`) and VS Code
 * Copilot Chat (`flavor: "Code"`) integrations. Adding a new vscode fork
 * (Insiders, Code-OSS, Windsurf, …) requires only extending the flavor union.
 *
 * Public symbols:
 *   - getVscodeUserDataDir(flavor, home?)
 *   - getVscodeWorkspaceStorageDir(flavor, home?)
 *   - findVscodeWorkspaceHash(flavor, projectDir)   — one repo → its workspace hash
 *   - listVscodeWorkspaceFolders(flavor)            — every workspace → its folder
 *   - normalizePathForMatch(p)
 *
 * The last two are the two directions of the same lookup, and which one a caller
 * wants follows from how many repos it serves: a per-repo question takes the find,
 * a machine-wide scan serving every registered repo takes the list. Calling the
 * find in a loop re-reads every `workspace.json` on the machine once per repo.
 */

import { readdir, readFile } from "node:fs/promises";
import { homedir, platform } from "node:os";
import { join } from "node:path";
import { fileURLToPath } from "node:url";
import { createLogger } from "../Logger.js";

const log = createLogger("VscodeWorkspaceLocator");

export type VscodeFlavor = "Cursor" | "Code" | "Code - Insiders" | "VSCodium" | "Windsurf";

/** All VS Code-family flavors Jolli scans for extension data. Directory name == flavor string. */
export const ALL_VSCODE_FLAVORS: ReadonlyArray<VscodeFlavor> = [
	"Code",
	"Code - Insiders",
	"Cursor",
	"VSCodium",
	"Windsurf",
];

/**
 * Returns the VS Code-family user-data root for the current platform.
 *
 *   darwin   ~/Library/Application Support/<flavor>
 *   linux    ~/.config/<flavor>
 *   win32    %APPDATA%/<flavor>  (fallback to ~/AppData/Roaming/<flavor>)
 */
export function getVscodeUserDataDir(flavor: VscodeFlavor, home: string = homedir()): string {
	switch (platform()) {
		case "darwin":
			return join(home, "Library", "Application Support", flavor);
		case "win32":
			return join(process.env.APPDATA ?? join(home, "AppData", "Roaming"), flavor);
		default:
			return join(home, ".config", flavor);
	}
}

/** Returns the workspaceStorage dir for the given flavor. */
export function getVscodeWorkspaceStorageDir(flavor: VscodeFlavor, home?: string): string {
	return join(getVscodeUserDataDir(flavor, home), "User", "workspaceStorage");
}

/**
 * Normalises a filesystem path for workspace matching.
 * - Converts backslashes to forward slashes so Windows paths from
 *   fileURLToPath (which returns `\`-separated paths) compare correctly
 *   against caller-supplied forward-slash paths.
 * - Strips trailing slashes (linear-time loop, not regex — avoids CodeQL polynomial-redos
 *   warnings on JSON-loaded paths).
 * - Lowercases on case-insensitive platforms (darwin, win32).
 */
export function normalizePathForMatch(p: string): string {
	const fwd = p.replace(/\\/g, "/");
	let end = fwd.length;
	while (end > 0 && fwd[end - 1] === "/") {
		end--;
	}
	const trimmed = fwd.slice(0, end);
	const os = platform();
	return os === "darwin" || os === "win32" ? trimmed.toLowerCase() : trimmed;
}

/**
 * Reads one `workspaceStorage/<entry>/workspace.json` and returns the local path its
 * `folder` URI points at, or undefined when there is none to read.
 *
 * Single-folder workspaces only — an entry with a `workspace` field instead of
 * `folder` (a multi-root `.code-workspace` file) yields undefined and is skipped by
 * every caller.
 */
async function readWorkspaceFolderPath(
	flavor: VscodeFlavor,
	wsStorageDir: string,
	entry: string,
): Promise<string | undefined> {
	const wsJsonPath = join(wsStorageDir, entry, "workspace.json");
	let folderUri: string | undefined;
	try {
		const raw = await readFile(wsJsonPath, "utf8");
		const parsed = JSON.parse(raw) as Record<string, unknown>;
		folderUri = typeof parsed.folder === "string" ? parsed.folder : undefined;
	} catch {
		return undefined;
	}

	if (!folderUri || !folderUri.startsWith("file://")) {
		return undefined;
	}

	try {
		return fileURLToPath(folderUri);
	} catch {
		log.warn("%s workspace %s has unparseable folder URI: %s", flavor, entry, folderUri);
		return undefined;
	}
}

/** One VS Code-family workspace: its storage directory name and the folder it opens. */
export interface VscodeWorkspaceFolder {
	/** The `workspaceStorage/<hash>` entry name. */
	readonly hash: string;
	/** The local path its `folder` URI resolves to. Raw — not normalised for matching. */
	readonly folderPath: string;
}

/**
 * Every single-folder workspace this flavour has storage for.
 *
 * The forward direction of {@link findVscodeWorkspaceHash}, and the one a machine-wide
 * scan needs. Resolving a hash FROM a repo path can only answer for the repo you
 * already have, so a scan built on it re-reads every `workspace.json` on the machine
 * once per registered repo. Listing them once gives every repo its answer from one
 * pass.
 *
 * Returns an empty list when the storage directory is unreadable — the same silent
 * degradation `findVscodeWorkspaceHash` applies, and for the same reason: a machine
 * without this editor installed is not a failure to report.
 */
export async function listVscodeWorkspaceFolders(flavor: VscodeFlavor): Promise<ReadonlyArray<VscodeWorkspaceFolder>> {
	const wsStorageDir = getVscodeWorkspaceStorageDir(flavor);

	let entries: string[];
	try {
		entries = await readdir(wsStorageDir);
	} catch {
		log.debug("%s workspaceStorage not readable at %s", flavor, wsStorageDir);
		return [];
	}

	const out: VscodeWorkspaceFolder[] = [];
	for (const entry of entries) {
		const folderPath = await readWorkspaceFolderPath(flavor, wsStorageDir, entry);
		if (folderPath !== undefined) out.push({ hash: entry, folderPath });
	}
	return out;
}

/**
 * Scans the workspaceStorage directory for an entry whose `workspace.json` has
 * a `folder` URI that resolves to projectDir. Returns the entry name (workspace
 * hash) on match, or null when no match is found.
 *
 * Single-folder workspaces only — entries with a `workspace` field instead of
 * `folder` (multi-root .code-workspace files) are skipped silently.
 *
 * Deliberately NOT written as a filter over {@link listVscodeWorkspaceFolders}: this
 * stops at the first match, and its callers ask about one repo at a time. A caller
 * that needs every workspace should use the list directly rather than calling this in
 * a loop.
 */
export async function findVscodeWorkspaceHash(flavor: VscodeFlavor, projectDir: string): Promise<string | null> {
	const wsStorageDir = getVscodeWorkspaceStorageDir(flavor);

	let entries: string[];
	try {
		entries = await readdir(wsStorageDir);
	} catch {
		log.debug("%s workspaceStorage not readable at %s", flavor, wsStorageDir);
		return null;
	}

	const target = normalizePathForMatch(projectDir);

	for (const entry of entries) {
		const folderPath = await readWorkspaceFolderPath(flavor, wsStorageDir, entry);
		if (folderPath !== undefined && normalizePathForMatch(folderPath) === target) {
			return entry;
		}
	}

	return null;
}

File: intellij/src/main/kotlin/ai/jolli/jollimemory/core/VscodeWorkspaceLocator.kt
package ai.jolli.jollimemory.core

import com.google.gson.JsonParser
import java.io.File
import java.net.URI
import java.nio.file.Paths

/**
 * VscodeWorkspaceLocator — per-platform path resolution and `workspace.json`
 * scanning for VS Code-family user data directories.
 *
 * Used by both Cursor IDE (`flavor: Cursor`) and VS Code Copilot Chat
 * (`flavor: Code`). Adding a new vscode fork (Insiders, Code-OSS, Windsurf, …)
 * requires only extending [VscodeFlavor].
 *
 * Kotlin port of `cli/src/core/VscodeWorkspaceLocator.ts`.
 */

private val log = JmLogger.create("VscodeWorkspaceLocator")

enum class VscodeFlavor(val dirName: String) {
	Cursor("Cursor"),
	Code("Code"),
}

/**
 * Returns the VS Code-family user-data root for the current platform.
 *
 *   darwin   ~/Library/Application Support/<flavor>
 *   linux    ~/.config/<flavor>
 *   win32    %APPDATA%/<flavor>  (fallback to ~/AppData/Roaming/<flavor>)
 *
 * All platform lookups (home dir, OS name, %APPDATA%) come from [env] so tests
 * can exercise every platform branch on any host without mutating JVM globals.
 */
fun getVscodeUserDataDir(flavor: VscodeFlavor, env: HookEnv = HookEnv()): String {
	val osName = env.osName.lowercase()
	val home = env.userHome.path
	return when {
		osName.contains("mac") ->
			home + File.separator + "Library" + File.separator + "Application Support" + File.separator + flavor.dirName
		osName.contains("win") ->
			(env.getenv("APPDATA") ?: (home + File.separator + "AppData" + File.separator + "Roaming")) +
				File.separator + flavor.dirName
		else ->
			home + File.separator + ".config" + File.separator + flavor.dirName
	}
}

/** Returns the workspaceStorage dir for the given flavor. */
fun getVscodeWorkspaceStorageDir(flavor: VscodeFlavor, env: HookEnv = HookEnv()): String =
	getVscodeUserDataDir(flavor, env) + File.separator + "User" + File.separator + "workspaceStorage"

/**
 * Normalises a filesystem path for workspace matching:
 *   - Backslashes → forward slashes (Windows path comparison)
 *   - Strip trailing slashes
 *   - Lowercase on macOS and Windows (case-insensitive filesystems)
 */
fun normalizePathForMatch(p: String, env: HookEnv = HookEnv()): String {
	val fwd = p.replace('\\', '/')
	val trimmed = fwd.trimEnd('/')
	val osName = env.osName.lowercase()
	return if (osName.contains("mac") || osName.contains("win")) trimmed.lowercase() else trimmed
}

/**
 * Scans the workspaceStorage directory for an entry whose `workspace.json` has
 * a `folder` URI that resolves to projectDir. Returns the entry name (workspace
 * hash) on match, or null when no match is found.
 *
 * Single-folder workspaces only — entries with a `workspace` field instead of
 * `folder` (multi-root .code-workspace files) are skipped silently.
 */
fun findVscodeWorkspaceHash(flavor: VscodeFlavor, projectDir: String, env: HookEnv = HookEnv()): String? {
	val wsStorageDir = File(getVscodeWorkspaceStorageDir(flavor, env))
	if (!wsStorageDir.isDirectory) {
		log.debug("%s workspaceStorage not readable at %s", flavor, wsStorageDir.path)
		return null
	}

	val target = normalizePathForMatch(projectDir, env)
	val entries = wsStorageDir.listFiles() ?: return null

	for (entry in entries) {
		val wsJson = File(entry, "workspace.json")
		if (!wsJson.isFile) continue

		val folderUri = try {
			JsonParser.parseString(wsJson.readText()).asJsonObject
				.get("folder")?.takeIf { it.isJsonPrimitive }?.asString
		} catch (_: Exception) {
			continue
		} ?: continue

		if (!folderUri.startsWith("file://")) continue

		val folderPath = try {
			Paths.get(URI(folderUri)).toString()
		} catch (_: Exception) {
			log.warn("%s workspace %s has unparseable folder URI: %s", flavor, entry.name, folderUri)
			continue
		}

		if (normalizePathForMatch(folderPath, env) == target) {
			return entry.name
		}
	}
	return null
}
